File: src/main/java/com/sinnerschrader/s2b/accounttool/logic/component/mapping/UserMapping.java
package com.sinnerschrader.s2b.accounttool.logic.component.mapping;

import com.sinnerschrader.s2b.accounttool.config.ldap.LdapConfiguration;
import com.sinnerschrader.s2b.accounttool.logic.entity.User;
import com.unboundid.ldap.sdk.SearchResultEntry;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.util.CollectionUtils;

import java.time.DateTimeException;
import java.time.LocalDate;
import java.util.Arrays;
import java.util.Map;

/**
 * Converter for Users from LDAP to Application
 */
public class UserMapping implements ModelMaping<User> {

	private static final Logger log = LoggerFactory.getLogger(UserMapping.class);

	@Autowired
	private LdapConfiguration ldapConfiguration;

	@Override
	public User map(SearchResultEntry entry) {
		if (entry == null) return null;

		Map.Entry<String, String> company = getCompany(entry.getDN(), entry.getAttributeValue("o"));
		final String dn = entry.getDN();
		final LocalDate birthDate =
				parseDate(
						dn,
						false,
						1972,
						entry.getAttributeValueAsInteger("szzBirthMonth"),
						entry.getAttributeValueAsInteger("szzBirthDay"));
		final LocalDate entryDate =
				parseDate(
						dn,
						true,
						entry.getAttributeValueAsInteger("szzEntryYear"),
						entry.getAttributeValueAsInteger("szzEntryMonth"),
						entry.getAttributeValueAsInteger("szzEntryDay"));
		final LocalDate exitDate =
				parseDate(
						dn,
						true,
						entry.getAttributeValueAsInteger("szzExitYear"),
						entry.getAttributeValueAsInteger("szzExitMonth"),
						entry.getAttributeValueAsInteger("szzExitDay"));

		return new User(
				dn,
				entry.getAttributeValue("uid"),
				entry.getAttributeValueAsInteger("uidNumber"),
				entry.getAttributeValueAsInteger("gidNumber"),
				entry.getAttributeValue("displayName"),
				entry.getAttributeValue("gecos"),
				entry.getAttributeValue("cn"),
				entry.getAttributeValue("givenName"),
				entry.getAttributeValue("sn"),
				entry.getAttributeValue("homeDirectory"),
				entry.getAttributeValue("loginShell"),
				birthDate,
				entry.getAttributeValue("sambaSID"),
				entry.getAttributeValue("sambaPasswordHistory"),
				entry.getAttributeValue("sambaAcctFlags"),
				entry.getAttributeValue("mail"),
				User.State.fromString(entry.getAttributeValue("szzStatus")),
				User.State.fromString(entry.getAttributeValue("szzMailStatus")),
				entry.getAttributeValueAsLong("sambaPwdLastSet"),
				entryDate,
				exitDate,
				entry.getAttributeValue("ou"),
				entry.getAttributeValue("description"),
				entry.getAttributeValue("telephoneNumber"),
				entry.getAttributeValue("mobile"),
				entry.getAttributeValue("employeeNumber"),
				entry.getAttributeValue("title"),
				entry.getAttributeValue("l"),
				entry.getAttributeValue("szzPublicKey"),
				company.getValue(),
				company.getKey());
	}

	private Map.Entry<String, String> getCompany(String dn, String organization) {
		for (Map.Entry<String, String> entry : ldapConfiguration.getCompaniesAsMap().entrySet()) {
			if (StringUtils.equals(entry.getValue(), organization)) {
				return entry;
			}
		}
		return extractCompanyFromDn(dn);
	}

	private LocalDate parseDate(
			String dn, boolean required, Integer year, Integer month, Integer day) {
		try {
			return LocalDate.of(year, month, day);
		} catch (DateTimeException dte) {
			log.error("Could not parse date on account " + dn, dte);
		} catch (Exception e) {
			if (required) {
				log.error("Date seems to be uncomplete on account " + dn, e);
			}
		}
		return null;
	}

	@Deprecated
	private Map.Entry<String, String> extractCompanyFromDn(String dn) {
		String[] parts = StringUtils.split(StringUtils.trimToEmpty(dn), ',');
		String key = "";
		if (parts.length == 5) {
			key = parts[2];
		}
		final String companyPrefix = "ou=";
		if (StringUtils.startsWith(key, companyPrefix)) {
			key = StringUtils.replaceOnce(key, companyPrefix, "");
		}
		if (ldapConfiguration.getCompaniesAsMap().containsKey(key)) {
			final String comKey = key;
			final String comVal = ldapConfiguration.getCompaniesAsMap().get(key);
			return new Map.Entry<String, String>() {

				@Override
				public String getKey() {
					return comKey;
				}

				@Override
				public String getValue() {
					return comVal;
				}

				@Override
				public String setValue(String value) {
					throw new IllegalAccessError("You cant set a new value on immutable Object.");
				}
			};
		}
		return null;
	}

	@Override
	public boolean isCompatible(SearchResultEntry entry) {
		return (entry != null)
				&& CollectionUtils.containsAny(
				Arrays.asList(entry.getObjectClassValues()), User.objectClasses);
	}
}

File: src/main/kotlin/com/sinnerschrader/s2b/accounttool/logic/component/mapping/UserMapping.kt
package com.sinnerschrader.s2b.accounttool.logic.component.mapping

import com.sinnerschrader.s2b.accounttool.config.UserConfiguration
import com.sinnerschrader.s2b.accounttool.config.ldap.LdapConfiguration
import com.sinnerschrader.s2b.accounttool.logic.entity.User
import com.unboundid.ldap.sdk.SearchResultEntry
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.*

@Service
class UserMapping {

    @Autowired
    private lateinit var ldapConfiguration: LdapConfiguration

    @Autowired
    private lateinit var userConfiguration: UserConfiguration

    fun map(entry: SearchResultEntry?): User? {
        if (entry == null) return null

        fun SearchResultEntry.int(attributeName: String) = getAttributeValueAsInteger(attributeName)
        fun SearchResultEntry.long(attributeName: String) = getAttributeValueAsLong(attributeName)
        fun SearchResultEntry.str(attributeName: String) = getAttributeValue(attributeName)
        fun SearchResultEntry.strMap(attributeName: String) =
                if (hasAttribute(attributeName)) str(attributeName).split(',')
                    .map { it.split('=', limit = 2).let { valuePair -> valuePair[0] to valuePair[1] } }.toMap()
                else emptyMap()

        return try {
            with(entry) {
                User(
                        dn = dn,
                        uid = str("uid"),
                        uidNumber = int("uidNumber"),
                        gidNumber = int("gidNumber"),
                        givenName = str("givenName"),
                        sn = str("sn"),
                        displayName = str("displayName"),
                        gecos = str("gecos"),
                        cn = str("cn"),
                        homeDirectory = str("homeDirectory"),
                        loginShell = str("loginShell"),
                        szzBirthDay = int("szzBirthDay"),
                        szzBirthMonth = int("szzBirthMonth"),
                        sambaSID = str("sambaSID"),
                        sambaPasswordHistory = str("sambaPasswordHistory"),
                        sambaAcctFlags = str("sambaAcctFlags"),
                        mail = str("mail"),
                        szzStatus = User.State.fromString(str("szzStatus")),
                        szzMailStatus = User.State.fromString(str("szzMailStatus")),
                        sambaPwdLastSet = long("sambaPwdLastSet") ?: 0L,
                        szzEntryDate = parseDate(dn, str("szzEntryDate")),
                        szzExitDate = parseDate(dn, str("szzExitDate")),
                        ou = str("ou"),
                        description = str("description"),
                        telephoneNumber = str("telephoneNumber") ?: "",
                        mobile = str("mobile") ?: "",
                        employeeNumber = str("employeeNumber"),
                        title = str("title") ?: "",
                        l = str("l"),
                        szzPublicKey = str("szzPublicKey") ?: "",
                        szzExternalAccounts = userConfiguration.externalAccounts.keys.map { it to "" }.toMap() + strMap("szzExternalAccounts"),
                        o = companyForDn(dn).second,
                        companyKey = companyForDn(dn).first,
                        modifiersName = str("modifiersName"),
                        modifytimestamp = str("modifytimestamp")
                )
            }
        } catch (e: Exception) {
            LOG.error("failed to map: " + entry.dn, e)
            null
        }
    }

    private fun companyForDn(dn: String) =
            try {
                with(Regex(",ou=([^,]+)").findAll(dn).last().groupValues[1]) {
                    this to (ldapConfiguration.companies[this] ?: "UNKNOWN")
                }
            } catch (e: NoSuchElementException) {
                "UNKNOWN" to "UNKNOWN"
            }


    private fun parseDate(dn: String, date: String?, required: Boolean = true): LocalDate? =
            date?.let {
                try {
                    LocalDate.parse(it)
                } catch (e: Exception) {
                    if (required)
                        LOG.warn("Could not parse date [dn=$dn, date=$date]")
                    null
                }
            }

    private fun parseDate(dn: String, required: Boolean, year: Int?, month: Int?, day: Int?): LocalDate? =
            try {
                LocalDate.of(year!!, month!!, day!!)
            } catch (e: Exception) {
                if (required)
                    LOG.warn("Could not parse date [dn=$dn, required=$required, year=$year, month=$month, day=$day]")
                null
            }

    companion object {
        private val LOG = LoggerFactory.getLogger(UserMapping::class.java)
    }
}
